File: quantum-sensing-jl/cfi.jl
# Classical Fisher Information for various measurement types.

export calculate_cfi, calculate_cfi_photon_counting, calculate_cfi_homodyne,
       calculate_cfi_heterodyne, calculate_cfi_sld_optimal, calculate_cfi_projective,
       compare_cfi

const PREC_DIFF_CFI = 1e-5
const PROB_FLOOR = 1e-15

# ── Internal helpers ──────────────────────────────────────────────────────

function _get_rho_drho(; dynamics::Function, param_value::Real=0.0,
                        param_type::Symbol=:Δ, prec::Real=PREC_DIFF_CFI, kwargs...)
    params_plus  = merge(kwargs, Dict(param_type => param_value + prec))
    params_minus = merge(kwargs, Dict(param_type => param_value - prec))
    params_center = merge(kwargs, Dict(param_type => param_value))

    function get_dense_dm(params)
        state = dynamics(; params...)
        if isket(state)
            state = ket2dm(state)
        end
        return Matrix{ComplexF64}(state.data)
    end

    ρ_plus  = get_dense_dm(params_plus)
    ρ_minus = get_dense_dm(params_minus)
    ρ_mat   = get_dense_dm(params_center)
    dρ_mat  = (ρ_plus - ρ_minus) / (2 * prec)

    return ρ_mat, dρ_mat
end

function _cfi_from_probs(probs::AbstractVector{<:Real},
                         dprobs::AbstractVector{<:Real};
                         prob_floor::Real=PROB_FLOOR)
    cfi = 0.0
    @inbounds for i in eachindex(probs)
        p = probs[i]
        if p > prob_floor
            cfi += dprobs[i]^2 / p
        end
    end
    return cfi
end

function _coherent_state_vec!(v::Vector{ComplexF64}, alpha::ComplexF64)
    N = length(v)
    v[1] = exp(-abs2(alpha) / 2)
    for n in 1:(N-1)
        v[n+1] = v[n] * alpha / sqrt(n)
    end
    return v
end

# ── Measurement-specific CFI functions ────────────────────────────────────

function calculate_cfi_photon_counting(rho::AbstractMatrix{<:Number},
                                       drho::AbstractMatrix{<:Number})
    N = size(rho, 1)
    probs  = Float64[real(rho[m, m]) for m in 1:N]
    dprobs = Float64[real(drho[m, m]) for m in 1:N]
    cfi = _cfi_from_probs(probs, dprobs)
    return (cfi=cfi, probabilities=probs, dprobabilities=dprobs,
            measurement_type=:photon_counting)
end

function calculate_cfi_homodyne(rho::AbstractMatrix{<:Number},
                                drho::AbstractMatrix{<:Number};
                                theta_LO::Real=0.0)
    N_basis = size(rho, 1)
    a = destroy(N_basis)
    X_theta = (a * exp(-im * theta_LO) + a' * exp(im * theta_LO)) / sqrt(2)
    evals, evecs = eigen(Hermitian(Matrix(X_theta.data)))

    probs  = Float64[real(dot(evecs[:, m], rho  * evecs[:, m])) for m in 1:N_basis]
    dprobs = Float64[real(dot(evecs[:, m], drho * evecs[:, m])) for m in 1:N_basis]
    probs .= max.(probs, 0.0)

    cfi = _cfi_from_probs(probs, dprobs)
    return (cfi=cfi, probabilities=probs, dprobabilities=dprobs,
            theta_LO=theta_LO, measurement_type=:homodyne)
end

function calculate_cfi_sld_optimal(rho::AbstractMatrix{<:Number},
                                    drho::AbstractMatrix{<:Number};
                                    sld_eigenvectors::Union{Nothing,AbstractMatrix}=nothing,
                                    prec_eig::Real=1e-8)
    N = size(rho, 1)

    if sld_eigenvectors === nothing
        evals_rho, evecs_rho = eigen(Hermitian(Matrix(rho)))
        dρ_rot = evecs_rho' * drho * evecs_rho
        L_rot = zeros(ComplexF64, N, N)
        for i in 1:N, j in 1:N
            denom = evals_rho[i] + evals_rho[j]
            if abs(denom) > prec_eig
                L_rot[i, j] = 2.0 * dρ_rot[i, j] / denom
            end
        end
        L_mat = evecs_rho * L_rot * evecs_rho'
        L_mat = (L_mat + L_mat') / 2.0
        qfi = real(tr(rho * L_mat^2))
        sld_evals, sld_evecs = eigen(Hermitian(L_mat))
    else
        sld_evecs = sld_eigenvectors
        sld_evals = Float64[real(dot(sld_evecs[:, m], drho * sld_evecs[:, m])) /
                            max(real(dot(sld_evecs[:, m], rho * sld_evecs[:, m])), PROB_FLOOR)
                            for m in 1:N]
        evals_rho, evecs_rho = eigen(Hermitian(Matrix(rho)))
        dρ_rot = evecs_rho' * drho * evecs_rho
        qfi = 0.0
        for i in 1:N, j in 1:N
            denom = evals_rho[i] + evals_rho[j]
            if abs(denom) > prec_eig
                qfi += 2.0 * abs2(dρ_rot[i, j]) / denom
            end
        end
        qfi = real(qfi)
    end

    probs  = Float64[real(dot(sld_evecs[:, m], rho  * sld_evecs[:, m])) for m in 1:N]
    dprobs = Float64[real(dot(sld_evecs[:, m], drho * sld_evecs[:, m])) for m in 1:N]
    cfi = _cfi_from_probs(probs, dprobs)

    return (cfi=cfi, probabilities=probs, dprobabilities=dprobs,
            sld_eigenvalues=sld_evals, sld_eigenvectors=sld_evecs,
            qfi=qfi, measurement_type=:sld_optimal)
end

function calculate_cfi_projective(rho::AbstractMatrix{<:Number},
                                   drho::AbstractMatrix{<:Number},
                                   projectors::Vector{<:AbstractVector{<:Number}})
    M = length(projectors)
    probs  = Float64[real(dot(projectors[m], rho  * projectors[m])) for m in 1:M]
    dprobs = Float64[real(dot(projectors[m], drho * projectors[m])) for m in 1:M]
    cfi = _cfi_from_probs(probs, dprobs)
    return (cfi=cfi, probabilities=probs, dprobabilities=dprobs,
            measurement_type=:projective)
end

function calculate_cfi_heterodyne(rho::AbstractMatrix{<:Number},
                                   drho::AbstractMatrix{<:Number};
                                   n_grid::Int=60,
                                   alpha_max::Union{Real,Nothing}=nothing,
                                   N_target::Union{Real,Nothing}=nothing)
    N_basis = size(rho, 1)

    if alpha_max === nothing
        if N_target !== nothing
            alpha_max = max(3.0, 1.5 * sqrt(N_target + 1))
        else
            a = destroy(N_basis)
            n_op = Matrix(a' * a)
            n_bar = real(tr(rho * n_op))
            alpha_max = max(3.0, 1.5 * sqrt(n_bar + 3 * sqrt(max(n_bar, 0.0)) + 1))
        end
    end

    alpha_re = range(-alpha_max, alpha_max; length=n_grid)
    alpha_im = range(-alpha_max, alpha_max; length=n_grid)
    dA = step(alpha_re) * step(alpha_im)

    n_total = n_grid * n_grid
    raw_probs  = Vector{Float64}(undef, n_total)
    raw_dprobs = Vector{Float64}(undef, n_total)
    v   = Vector{ComplexF64}(undef, N_basis)
    tmp = Vector{ComplexF64}(undef, N_basis)

    k = 0
    for ai in alpha_im, ar in alpha_re
        k += 1
        α = ComplexF64(ar, ai)
        _coherent_state_vec!(v, α)
        mul!(tmp, rho, v)
        raw_probs[k] = max(real(dot(v, tmp)), 0.0)
        mul!(tmp, drho, v)
        raw_dprobs[k] = real(dot(v, tmp))
    end

    cfi = _cfi_from_probs(raw_probs, raw_dprobs) * dA

    normalization = sum(raw_probs) * dA / π
    if abs(normalization - 1.0) > 0.05
        @warn "Heterodyne grid normalization = $(round(normalization; digits=4)) " *
              "(expected ≈ 1.0). Consider increasing n_grid or alpha_max."
    end

    return (cfi=cfi, alpha_max=alpha_max, n_grid=n_grid,
            normalization=normalization, measurement_type=:heterodyne)
end

# ── Unified entry point ──────────────────────────────────────────────────

function calculate_cfi(; dynamics::Union{Function,Nothing}=nothing,
                        param_value::Real=0.0,
                        param_type::Symbol=:Δ,
                        prec::Real=PREC_DIFF_CFI,
                        rho::Union{Nothing,AbstractMatrix}=nothing,
                        drho::Union{Nothing,AbstractMatrix}=nothing,
                        measurement::Symbol=:photon_counting,
                        kwargs...)
    if rho === nothing || drho === nothing
        dynamics === nothing && error("Must provide either `dynamics` or both `rho` and `drho`")
        rho, drho = _get_rho_drho(; dynamics=dynamics, param_value=param_value,
                                    param_type=param_type, prec=prec, kwargs...)
    end

    if measurement == :photon_counting
        return calculate_cfi_photon_counting(rho, drho)
    elseif measurement == :homodyne
        theta_LO = get(kwargs, :theta_LO, 0.0)
        return calculate_cfi_homodyne(rho, drho; theta_LO=theta_LO)
    elseif measurement == :heterodyne
        n_grid = get(kwargs, :n_grid, 60)
        alpha_max = get(kwargs, :alpha_max, nothing)
        N_target = get(kwargs, :N_target, nothing)
        return calculate_cfi_heterodyne(rho, drho; n_grid=n_grid,
                                        alpha_max=alpha_max, N_target=N_target)
    elseif measurement == :sld_optimal
        sld_eigenvectors = get(kwargs, :sld_eigenvectors, nothing)
        return calculate_cfi_sld_optimal(rho, drho; sld_eigenvectors=sld_eigenvectors)
    elseif measurement == :projective
        projectors = kwargs[:projectors]
        return calculate_cfi_projective(rho, drho, projectors)
    else
        error("Unknown measurement type: $measurement. " *
              "Use :photon_counting, :homodyne, :heterodyne, :sld_optimal, or :projective")
    end
end

# ── Compare multiple measurements ────────────────────────────────────────

function compare_cfi(; dynamics::Union{Function,Nothing}=nothing,
                      rho::Union{Nothing,AbstractMatrix}=nothing,
                      drho::Union{Nothing,AbstractMatrix}=nothing,
                      param_value::Real=0.0,
                      param_type::Symbol=:Δ,
                      prec::Real=PREC_DIFF_CFI,
                      measurements::Vector{Symbol}=[:photon_counting, :homodyne, :sld_optimal],
                      homodyne_phases::Vector{<:Real}=[0.0],
                      kwargs...)
    if rho === nothing || drho === nothing
        dynamics === nothing && error("Must provide either `dynamics` or both `rho` and `drho`")
        rho, drho = _get_rho_drho(; dynamics=dynamics, param_value=param_value,
                                    param_type=param_type, prec=prec, kwargs...)
    end

    results = Dict{Symbol, Any}()
    sld_result = calculate_cfi_sld_optimal(rho, drho)
    qfi = sld_result.qfi

    for meas in measurements
        if meas == :photon_counting
            results[:photon_counting] = calculate_cfi_photon_counting(rho, drho)
        elseif meas == :homodyne
            best_hom = nothing
            for θ_LO in homodyne_phases
                res = calculate_cfi_homodyne(rho, drho; theta_LO=θ_LO)
                if best_hom === nothing || res.cfi > best_hom.cfi
                    best_hom = res
                end
            end
            results[:homodyne] = best_hom
        elseif meas == :heterodyne
            het_kwargs = Dict{Symbol,Any}()
            haskey(kwargs, :n_grid) && (het_kwargs[:n_grid] = kwargs[:n_grid])
            haskey(kwargs, :alpha_max) && (het_kwargs[:alpha_max] = kwargs[:alpha_max])
            haskey(kwargs, :N_target) && (het_kwargs[:N_target] = kwargs[:N_target])
            results[:heterodyne] = calculate_cfi_heterodyne(rho, drho; het_kwargs...)
        elseif meas == :sld_optimal
            results[:sld_optimal] = sld_result
        elseif meas == :projective
            if haskey(kwargs, :projectors)
                results[:projective] = calculate_cfi_projective(rho, drho, kwargs[:projectors])
            end
        end
    end

    best_meas = :sld_optimal
    best_cfi  = 0.0
    for (k, v) in results
        if v.cfi > best_cfi
            best_cfi  = v.cfi
            best_meas = k
        end
    end

    efficiency = Dict{Symbol, Float64}()
    for (k, v) in results
        efficiency[k] = qfi > 0 ? v.cfi / qfi : 0.0
    end

    return (qfi=qfi, results=results, best_measurement=best_meas,
            best_cfi=best_cfi, efficiency=efficiency)
end

File: quantum_sensing/cfi.py
"""
Classical Fisher Information for various measurement types.

Python port of julia/src/cfi.jl with identical physics and constants.

Every measurement reduces to the same formula on its outcome distribution:

    CFI = sum_m (d_theta p_m)^2 / p_m,   p_m = Tr[rho Pi_m],
                                         d_theta p_m = Tr[drho Pi_m]

with (rho, drho = d rho / d theta) either supplied directly (e.g. from a
cache) or built by central finite differences through a `dynamics` callable
(same pattern as :func:`quantum_sensing.sld.calculate_qfi`).

Measurement types
-----------------
photon_counting : Fock projectors (diagonal of rho, drho)
homodyne        : eigenbasis of X_theta = (a e^{-i theta} + a^dag e^{i theta})/sqrt(2)
heterodyne      : coherent-state POVM |alpha><alpha|/pi on a phase-space grid
sld_optimal     : eigenbasis of the SLD (saturates the QFI by construction)
projective      : arbitrary user-supplied projector vectors
"""

import warnings

import numpy as np
from numpy.linalg import eigh
import qutip as qt

PREC_DIFF_CFI = 1e-5
PROB_FLOOR = 1e-15


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------

def _get_rho_drho(dynamics, param_value=0.0, param_type="Delta",
                  prec=PREC_DIFF_CFI, **kwargs):
    """(rho, drho) by central finite differences through `dynamics`."""
    def dense_dm(params):
        state = dynamics(**params)
        if state.isket:
            v = state.full()
            return v @ v.conj().T
        return state.full()

    rho_plus = dense_dm({**kwargs, param_type: param_value + prec})
    rho_minus = dense_dm({**kwargs, param_type: param_value - prec})
    rho = dense_dm({**kwargs, param_type: param_value})
    drho = (rho_plus - rho_minus) / (2 * prec)
    return rho, drho


def _cfi_from_probs(probs, dprobs, prob_floor=PROB_FLOOR):
    """CFI = sum dp^2/p over outcomes with p > prob_floor."""
    probs = np.asarray(probs, dtype=float)
    dprobs = np.asarray(dprobs, dtype=float)
    mask = probs > prob_floor
    return float(np.sum(dprobs[mask] ** 2 / probs[mask]))


def _coherent_state_vec(alpha, N):
    """Fock-basis amplitudes of |alpha> up to N-1 (stable recurrence)."""
    v = np.empty(N, dtype=complex)
    v[0] = np.exp(-abs(alpha) ** 2 / 2)
    for n in range(1, N):
        v[n] = v[n - 1] * alpha / np.sqrt(n)
    return v


def _projector_probs(rho, drho, vecs):
    """p_m, dp_m for projectors onto the columns of `vecs`."""
    probs = np.real(np.einsum("im,ij,jm->m", vecs.conj(), rho, vecs, optimize=True))
    dprobs = np.real(np.einsum("im,ij,jm->m", vecs.conj(), drho, vecs, optimize=True))
    return probs, dprobs


# ---------------------------------------------------------------------------
# Measurement-specific CFI functions
# ---------------------------------------------------------------------------

def calculate_cfi_photon_counting(rho, drho):
    """CFI of ideal photon counting: p_m = rho_mm."""
    probs = np.real(np.diag(rho)).copy()
    dprobs = np.real(np.diag(drho)).copy()
    return {
        "cfi": _cfi_from_probs(probs, dprobs),
        "probabilities": probs,
        "dprobabilities": dprobs,
        "measurement_type": "photon_counting",
    }


def calculate_cfi_homodyne(rho, drho, theta_LO=0.0):
    """CFI of balanced homodyne at LO phase theta_LO, via the eigenbasis of
    the truncated quadrature operator X_theta = (a e^{-i th} + a^dag e^{i th})/sqrt(2).

    Convention: theta_LO = 0 measures X = (a + a^dag)/sqrt(2).
    """
    N = rho.shape[0]
    a = qt.destroy(N).full()
    X_theta = (a * np.exp(-1j * theta_LO) + a.conj().T * np.exp(1j * theta_LO)) / np.sqrt(2)
    _, evecs = eigh(X_theta)

    probs, dprobs = _projector_probs(rho, drho, evecs)
    probs = np.maximum(probs, 0.0)

    return {
        "cfi": _cfi_from_probs(probs, dprobs),
        "probabilities": probs,
        "dprobabilities": dprobs,
        "theta_LO": theta_LO,
        "measurement_type": "homodyne",
    }


def calculate_cfi_sld_optimal(rho, drho, sld_eigenvectors=None, prec_eig=1e-8):
    """CFI of the SLD-eigenbasis measurement (saturates the QFI).

    If `sld_eigenvectors` is given, evaluate that fixed basis instead
    (useful for testing a basis calibrated at a different channel point).
    Also returns the QFI.
    """
    N = rho.shape[0]
    evals_rho, evecs_rho = eigh(rho)
    drho_rot = evecs_rho.conj().T @ drho @ evecs_rho

    if sld_eigenvectors is None:
        denom = evals_rho[:, None] + evals_rho[None, :]
        L_rot = np.where(np.abs(denom) > prec_eig, 2.0 * drho_rot / np.where(denom == 0, 1, denom), 0.0)
        L = evecs_rho @ L_rot @ evecs_rho.conj().T
        L = (L + L.conj().T) / 2.0
        qfi = float(np.real(np.trace(rho @ L @ L)))
        sld_evals, sld_evecs = eigh(L)
    else:
        sld_evecs = sld_eigenvectors
        p_basis, dp_basis = _projector_probs(rho, drho, sld_evecs)
        sld_evals = dp_basis / np.maximum(p_basis, PROB_FLOOR)
        denom = evals_rho[:, None] + evals_rho[None, :]
        mask = np.abs(denom) > prec_eig
        qfi = float(np.real(np.sum(2.0 * np.abs(drho_rot[mask]) ** 2
                                   / denom[mask])))

    probs, dprobs = _projector_probs(rho, drho, sld_evecs)
    return {
        "cfi": _cfi_from_probs(probs, dprobs),
        "probabilities": probs,
        "dprobabilities": dprobs,
        "sld_eigenvalues": sld_evals,
        "sld_eigenvectors": sld_evecs,
        "qfi": qfi,
        "measurement_type": "sld_optimal",
    }


def calculate_cfi_projective(rho, drho, projectors):
    """CFI of an arbitrary projective measurement.

    `projectors` is a list of state vectors (1-D arrays) or a matrix whose
    columns are the projector vectors.
    """
    vecs = np.column_stack(projectors) if isinstance(projectors, (list, tuple)) \
        else np.asarray(projectors)
    probs, dprobs = _projector_probs(rho, drho, vecs)
    return {
        "cfi": _cfi_from_probs(probs, dprobs),
        "probabilities": probs,
        "dprobabilities": dprobs,
        "measurement_type": "projective",
    }


def calculate_cfi_heterodyne(rho, drho, n_grid=60, alpha_max=None, N_target=None):
    """CFI of ideal heterodyne (double homodyne): POVM |alpha><alpha|/pi on
    an n_grid x n_grid phase-space grid of half-width alpha_max.

    If alpha_max is None it is sized from N_target, or else from <n> of rho.
    Warns if the grid misses probability (normalization far from 1).
    """
    N = rho.shape[0]

    if alpha_max is None:
        if N_target is not None:
            alpha_max = max(3.0, 1.5 * np.sqrt(N_target + 1))
        else:
            n_bar = float(np.real(np.sum(np.arange(N) * np.real(np.diag(rho)))))
            alpha_max = max(3.0, 1.5 * np.sqrt(n_bar + 3 * np.sqrt(max(n_bar, 0.0)) + 1))

    alpha_re = np.linspace(-alpha_max, alpha_max, n_grid)
    alpha_im = np.linspace(-alpha_max, alpha_max, n_grid)
    dA = (alpha_re[1] - alpha_re[0]) * (alpha_im[1] - alpha_im[0])

    # coherent-state matrix: columns |alpha_k> for every grid point
    ar, ai = np.meshgrid(alpha_re, alpha_im)
    alphas = (ar + 1j * ai).ravel()
    ns = np.arange(N)
    # log-domain recurrence, vectorized: v_n = e^{-|a|^2/2} a^n / sqrt(n!)
    with np.errstate(divide="ignore"):
        log_fact = np.cumsum(np.log(np.maximum(ns, 1)))
    V = np.exp(-np.abs(alphas[None, :]) ** 2 / 2
               + ns[:, None] * np.log(np.where(alphas == 0, 1, alphas))[None, :]
               - log_fact[:, None] / 2)
    V[:, alphas == 0] = 0.0
    V[0, alphas == 0] = 1.0

    raw_probs = np.maximum(np.real(np.einsum("ik,ij,jk->k", V.conj(), rho, V,
                                             optimize=True)), 0.0)
    raw_dprobs = np.real(np.einsum("ik,ij,jk->k", V.conj(), drho, V, optimize=True))

    # F_C = int (dQ/dtheta)^2 / Q d^2alpha with Q(alpha) = <alpha|rho|alpha> / pi.
    # Per-cell prob is p = Q dA, so dp^2/p carries one net factor of 1/pi.
    # Matches julia/src/cfi.jl calculate_cfi_heterodyne.
    cfi = _cfi_from_probs(raw_probs, raw_dprobs) * dA / np.pi

    normalization = float(raw_probs.sum() * dA / np.pi)
    if abs(normalization - 1.0) > 0.05:
        warnings.warn(
            f"Heterodyne grid normalization = {normalization:.4f} "
            "(expected ~ 1.0). Consider increasing n_grid or alpha_max.")

    return {
        "cfi": cfi,
        "alpha_max": alpha_max,
        "n_grid": n_grid,
        "normalization": normalization,
        "measurement_type": "heterodyne",
    }


# ---------------------------------------------------------------------------
# Unified entry point
# ---------------------------------------------------------------------------

def calculate_cfi(dynamics=None, param_value=0.0, param_type="Delta",
                  prec=PREC_DIFF_CFI, rho=None, drho=None,
                  measurement="photon_counting", **kwargs):
    """Unified CFI: supply either (rho, drho) or a `dynamics` callable
    (finite-differenced in `param_type`, same pattern as calculate_qfi).

    measurement in {"photon_counting", "homodyne", "heterodyne",
                    "sld_optimal", "projective"}.
    Measurement options are passed through kwargs (theta_LO, n_grid,
    alpha_max, N_target, sld_eigenvectors, projectors); remaining kwargs go
    to `dynamics`.
    """
    meas_keys = {"theta_LO", "n_grid", "alpha_max", "N_target",
                 "sld_eigenvectors", "projectors"}
    meas_kwargs = {k: kwargs.pop(k) for k in list(kwargs) if k in meas_keys}

    if rho is None or drho is None:
        if dynamics is None:
            raise ValueError("Must provide either `dynamics` or both `rho` and `drho`")
        if rho is not None:
            # `rho` without `drho` is the *input* probe state for `dynamics`
            # (same meaning as in calculate_qfi), not a precomputed output.
            kwargs["rho"] = rho
        rho, drho = _get_rho_drho(dynamics, param_value=param_value,
                                  param_type=param_type, prec=prec, **kwargs)

    if measurement == "photon_counting":
        return calculate_cfi_photon_counting(rho, drho)
    elif measurement == "homodyne":
        return calculate_cfi_homodyne(rho, drho,
                                      theta_LO=meas_kwargs.get("theta_LO", 0.0))
    elif measurement == "heterodyne":
        return calculate_cfi_heterodyne(rho, drho,
                                        n_grid=meas_kwargs.get("n_grid", 60),
                                        alpha_max=meas_kwargs.get("alpha_max"),
                                        N_target=meas_kwargs.get("N_target"))
    elif measurement == "sld_optimal":
        return calculate_cfi_sld_optimal(
            rho, drho, sld_eigenvectors=meas_kwargs.get("sld_eigenvectors"))
    elif measurement == "projective":
        return calculate_cfi_projective(rho, drho, meas_kwargs["projectors"])
    else:
        raise ValueError(
            f"Unknown measurement type: {measurement}. Use 'photon_counting', "
            "'homodyne', 'heterodyne', 'sld_optimal', or 'projective'")


# ---------------------------------------------------------------------------
# Compare multiple measurements
# ---------------------------------------------------------------------------

def compare_cfi(dynamics=None, rho=None, drho=None, param_value=0.0,
                param_type="Delta", prec=PREC_DIFF_CFI,
                measurements=("photon_counting", "homodyne", "sld_optimal"),
                homodyne_phases=(0.0,), **kwargs):
    """CFI of several measurements on the same (rho, drho), plus the QFI and
    per-measurement efficiency CFI/QFI. Homodyne is maximized over
    `homodyne_phases`.
    """
    meas_keys = {"n_grid", "alpha_max", "N_target", "projectors"}
    meas_kwargs = {k: kwargs.pop(k) for k in list(kwargs) if k in meas_keys}

    if rho is None or drho is None:
        if dynamics is None:
            raise ValueError("Must provide either `dynamics` or both `rho` and `drho`")
        if rho is not None:
            # `rho` without `drho` is the *input* probe state for `dynamics`
            # (same meaning as in calculate_qfi), not a precomputed output.
            kwargs["rho"] = rho
        rho, drho = _get_rho_drho(dynamics, param_value=param_value,
                                  param_type=param_type, prec=prec, **kwargs)

    results = {}
    sld_result = calculate_cfi_sld_optimal(rho, drho)
    qfi = sld_result["qfi"]

    for meas in measurements:
        if meas == "photon_counting":
            results["photon_counting"] = calculate_cfi_photon_counting(rho, drho)
        elif meas == "homodyne":
            best = None
            for th in homodyne_phases:
                res = calculate_cfi_homodyne(rho, drho, theta_LO=th)
                if best is None or res["cfi"] > best["cfi"]:
                    best = res
            results["homodyne"] = best
        elif meas == "heterodyne":
            results["heterodyne"] = calculate_cfi_heterodyne(
                rho, drho,
                n_grid=meas_kwargs.get("n_grid", 60),
                alpha_max=meas_kwargs.get("alpha_max"),
                N_target=meas_kwargs.get("N_target"))
        elif meas == "sld_optimal":
            results["sld_optimal"] = sld_result
        elif meas == "projective":
            if "projectors" in meas_kwargs:
                results["projective"] = calculate_cfi_projective(
                    rho, drho, meas_kwargs["projectors"])

    best_meas, best_cfi = "sld_optimal", 0.0
    for k, v in results.items():
        if v["cfi"] > best_cfi:
            best_cfi, best_meas = v["cfi"], k

    efficiency = {k: (v["cfi"] / qfi if qfi > 0 else 0.0) for k, v in results.items()}

    return {
        "qfi": qfi,
        "results": results,
        "best_measurement": best_meas,
        "best_cfi": best_cfi,
        "efficiency": efficiency,
    }
